### PR TITLE
Use Firestore data when reseting prod task and rerunning a build

### PR DESCRIPTION
**Pull request by keyonghan**
Part of https://github.com/flutter/flutter/issues/142951

This PR:
1) supports `taskDocumentName` arg for reset-prod-task API. This is to prepare frontend logic switched to Firestore
2) uses Firestore data when checking rerun, instead of Datastore data
3) moves `taskFailStatusSet` to Task models
4) moves Firestore logics out of try/catch block to break API if failing

### CODE REVIEW

**Comment by ricardoamador** [comment]
Quick question, this does not change functionality does it?

File: app_dart/lib/src/service/luci_build_service.dart
Comment (ricardoamador) [inline app_dart/lib/src/service/luci_build_service.dart:29]: Ha! I did something similar in my branch!

File: app_dart/lib/src/service/luci_build_service.dart
Comment (keyonghan) [inline app_dart/lib/src/service/luci_build_service.dart:731]: This is the main change: use firestore task to determine whether rerun.

**Comment by keyonghan** [comment]
> Quick question, this does not change functionality does it?

Added a comment for the main change.

**Review by ricardoamador — APPROVED** [review]
Can we wait until Monday to merge this? I have some testing I would like to get in today before needing to refactor on top of this.

**Comment by keyonghan** [comment]
> Can we wait until Monday to merge this? I have some testing I would like to get in today before needing to refactor on top of this.

You got it. Will add the label Monday.

**Comment by ricardoamador** [comment]
> > Can we wait until Monday to merge this? I have some testing I would like to get in today before needing to refactor on top of this.
> 
> You got it. Will add the label Monday.

Awesome! Thank you!

**Comment by keyonghan** [comment]
Landing now.

**Pull request merged**
